File: src/R/sv_by_orf_generation.R
library(ggplot2) 
# Récupération du chemin du script
args <- commandArgs(trailingOnly = FALSE)
file_arg <- "--file="
script_path <- sub(file_arg, "", args[grep(file_arg, args)])
script_dir <- dirname(normalizePath(script_path))  # Dossier du script

# Fichier d'entrée : tableau des SV par ORF et génération 
infile <- file.path(script_dir, "../../Files/plot_files/sv_by_orf_generation_wide.csv")
# Fichier de sortie : heatmap
out_plot <- file.path(script_dir, "../../Plots/sv_by_orf_generation.png")
# Ordre des générations
gen_order <- c("P15", "P25", "P27", "P30", "P50", "P65")

# Nombre d'ORFs à afficher (top 30 ici)
top_n_orfs <- 30

# Lecture du CSV 
d <- read.csv(infile, stringsAsFactors = FALSE, check.names = FALSE)
# Sélection des ORFs les plus importants (top N)
d <- head(d, top_n_orfs)


plot_long <- reshape(
  d[, c("orf", gen_order)],  # colonnes utilisées
  varying = gen_order,       # colonnes à transformer
  v.names = "n_unique_sv",   # nom de la variable créée
  timevar = "generation",    # nom de la variable génération
  times = gen_order,         # valeurs de génération
  direction = "long"         # transformation vers format long
)

# Conversion en facteur ordonné pour garder l'ordre des générations
plot_long$generation <- factor(plot_long$generation, levels = gen_order, ordered = TRUE)
# Ordre des ORFs inversé pour affichage (du plus important en haut)
plot_long$orf <- factor(plot_long$orf, levels = rev(d$orf))

# Création de la heatmap
p <- ggplot(plot_long, aes(x = generation, y = orf, fill = n_unique_sv)) +
  geom_tile(color = "white", linewidth = 0.3) +  # cases de la heatmap
  geom_text(aes(label = n_unique_sv), size = 2.5) +  # valeurs affichées dans les cases
  scale_fill_gradient(low = "#fff7bc", high = "#d95f0e") +  # couleur foncé = bcp de variants et clair - peu de variants
  labs(
    title = paste("Top", nrow(d), "ORF par nombre de SV uniques et par generation"),
    x = "Generation",
    y = "ORF",
    fill = "Nombre de SV uniques"
  ) +
  theme_minimal(base_size = 12) +
  theme(
    panel.grid = element_blank(),  
    axis.text.y = element_text(size = 8)  
  )
height <- max(8, 0.3 * nrow(d) + 2)

# Sauvegarde du graphique
ggsave(out_plot, p, width = 9, height = height, dpi = 300)

# Message de confirmation
cat("Sauvegardé:", out_plot, "\n")

File: src/Python/sv_by_orf_generation.py
#!/usr/bin/env python3
from pathlib import Path 
import pandas as pd  


# Définition des chemins de dossier du script et racine du projet
SCRIPT_DIR = Path(__file__).resolve().parent  
PROJECT_ROOT = SCRIPT_DIR.parent.parent  

# Fichier d'entrée contenant les variants SV annotés
INPUT_CSV = PROJECT_ROOT / "Files" / "plot_files" / "summary_table_sv_with_ORF.csv"
# Fichier de sortie 1
OUT_COUNTS_LONG = PROJECT_ROOT / "Files" / "plot_files" / "sv_by_orf_generation_summary.csv"
# Fichier de sortie 2
OUT_COUNTS_WIDE = PROJECT_ROOT / "Files" / "plot_files" / "sv_by_orf_generation_wide.csv"
# Ordre des générations
GEN_ORDER = ["P15", "P25", "P27", "P30", "P50", "P65"]


# Fonction pour séparer les ORFs multiples
def split_orfs(value: object) -> list[str]:
    if pd.isna(value) or str(value).strip() == "":
        return []  # aucun ORF
    # séparation par ";" et nettoyage des espaces
    return [part.strip() for part in str(value).split(";") if part.strip()]


def main() -> None:
    # Lecture du fichier CSV
    df = pd.read_csv(INPUT_CSV)
    # Conversion d'echantillon en numérique
    df["replicate"] = pd.to_numeric(df["replicate"], errors="coerce")
    # Filtrage des echantillons valides (1 à 10)
    df = df[df["replicate"].between(1, 10)].copy()
    # Filtrage des générations d'intérêt
    df = df[df["generation"].isin(GEN_ORDER)].copy()
    # Création d'un identifiant unique pour chaque variant
    df["variant_id"] = df[["chrom", "position", "svtype", "ref", "alt", "svlen"]].astype(str).agg("|".join, axis=1)
    # Séparation des ORFs multiples en listes
    df["orf"] = df["orf"].apply(split_orfs)
    # une ligne par ORF
    df = df.explode("orf")
    # Suppression des ORFs vides
    df = df[df["orf"].notna() & (df["orf"] != "")].copy()
    # Suppression des variants intergéniques
    df = df[df["orf"] != "Intergenic"].copy()

    # Comptage des variants uniques par ORF et génération
    counts = (
        df.groupby(["orf", "generation"], as_index=False)["variant_id"]
        .nunique()  # nombre de variants uniques
        .rename(columns={"variant_id": "n_unique_sv"})
    )

    # Conversion de la génération en facteur ordonné
    counts["generation"] = pd.Categorical(
        counts["generation"],
        categories=GEN_ORDER,
        ordered=True
    )

    # Tri des résultats
    counts = counts.sort_values(["orf", "generation"])

    # Sauvegarde au format long
    counts.to_csv(OUT_COUNTS_LONG, index=False)

    
    # Transformation en format large pour visualiser
    wide = (
        counts.pivot(index="orf", columns="generation", values="n_unique_sv")
        .reindex(columns=GEN_ORDER)  # garantir l'ordre des générations
        .fillna(0)  # remplacer les valeurs manquantes par 0
        .astype(int)
    )

    # Calcul du total de SV par ORF
    wide["total_sv"] = wide.sum(axis=1)
    # Tri décroissant selon le total
    wide = wide.sort_values("total_sv", ascending=False)
    # Sauvegarde au format large
    wide.to_csv(OUT_COUNTS_WIDE)

    # Messages de confirmation
    print("Sauvegardé:")
    print("-", str(OUT_COUNTS_LONG))
    print("-", str(OUT_COUNTS_WIDE))

if __name__ == "__main__":
    main()